versiondb/tsrocksdb/store.go

```diff
@@ -216,6 +232,62 @@ func (s Store) Flush() error {
 	)
 }
 
+// FixData fixes wrong data written in versiondb due to rocksdb upgrade, the operation is idempotent.
+// see: https://github.com/crypto-org-chain/cronos/issues/1683
+// call this before `SetSkipVersionZero(true)`.
+func (s Store) FixData(storeKeys []types.StoreKey) error {
+	for _, storeKey := range storeKeys {
+		if err := s.fixDataStore(storeKey.Name()); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// fixDataStore iterate the wrong data at version 0, parse the timestamp from the key and write it again.
+func (s Store) fixDataStore(storeName string) error {
+	var version int64
+	iter, err := s.IteratorAtVersion(storeName, nil, nil, &version)
+	if err != nil {
+		return err
+	}
+	defer iter.Close()
+
+	batch := grocksdb.NewWriteBatch()
+	defer batch.Destroy()
+
+	prefix := storePrefix(storeName)
+	for ; iter.Valid(); iter.Next() {
+		key := iter.Key()
+		if len(key) < TimestampSize {
+			return fmt.Errorf("invalid key length: %X", key)
+		}
+
+		ts := key[len(key)-TimestampSize:]
+		key = key[:len(key)-TimestampSize]
+		realKey := cloneAppend(prefix, key)
+
+		readOpts := grocksdb.NewDefaultReadOptions()
+		readOpts.SetTimestamp(ts)
+		oldValue, err := s.db.GetCF(readOpts, s.cfHandle, realKey)
```

Comment:
can oldValue be nil without error?

Reply:
no, but it can be an empty slice, in that case `Data` returns `nil` slice.